### PR TITLE
feat!: reduce sorted runs during compaction

### PR DESCRIPTION
**Pull request by v0y4g3r**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## Refer to a related PR or issue link (optional)
- #3416

## What's changed and what's your intention?

This PR changes the behavior during compaction from reduce file nums to sorted runs.

A sorted run in GreptimeDB is defined as a sequence of non-overlapping SST files. If multiple sorted runs exist in some time window, even a simple query with explicit time ranges may invole multiple files. Before this PR, GreptimeDB merges all files in active window into one single file if the total num of files exceeds `max_active_window_files`, of which the write amplification is rather severe. In this PR, we only enforce a `max_active_window_runs` in active window to reduce read amplication. If num of sorted runs exceeds that threshold, it only merges the overlapping files on demand.

## Checklist

- [x]  I have written the necessary rustdoc comments.
- [x]  I have added the necessary unit tests and integration tests.
- [x]  This PR requires documentation updates.

## Summary by CodeRabbit

- **New Features**
  - Added the ability to merge ranged items efficiently and manage sorted runs in the new `run` module.

- **Improvements**
  - Renamed compaction parameters from `max_active_window_files` and `max_inactive_window_files` to `max_active_window_runs` and `max_inactive_window_runs` for clarity.
  - Updated logic in compaction-related modules to handle runs instead of files.

- **Bug Fixes**
  - Corrected configurations and assertions in tests to align with the updated compaction parameters.

### CODE REVIEW

**Review by evenyag — APPROVED** [review]

**Review by waynexia — APPROVED** [review]
Rest LGTM

**Comment by killme2008** [comment]
What's the progress of this PR? @v0y4g3r

**Comment by killme2008** [comment]
@v0y4g3r  What's the status of this PR?  It's deprecated or just wait for resolving conflicts.

File: src/mito2/src/region/options.rs [hunk 1]
```diff
@@ -130,9 +130,9 @@ impl Default for CompactionOptions {
 #[derive(Debug, Clone, PartialEq, Eq, Deserialize)]
 #[serde(default)]
 pub struct TwcsOptions {
-    /// Max num of files that can be kept in active writing time window.
+    /// Max num of sorted runs that can be kept in active writing time window.
     #[serde_as(as = "DisplayFromStr")]
-    pub max_active_window_files: usize,
+    pub max_active_window_runs: usize,
```
[the remark this answers is not in the document]
Comment (evenyag) [reply src/mito2/src/region/options.rs:180]: Marked as breaking change. Most users don't set this option so it should be fine.

**Pull request merged**